File: modules/hinsdale-engine/android/src/main/java/expo/modules/hinsdaleengine/HinsdaleEngineModule.kt
package expo.modules.hinsdaleengine

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.Executors

class HinsdaleEngineModule : Module() {
  private val engineExecutor = Executors.newSingleThreadExecutor { runnable -> Thread(runnable, "hinsdale-engine") }

  companion object { init { System.loadLibrary("hinsdale_bridge") } }

  private external fun nativeAnalyze(bytecodeHex: String, qualityTier: String): String
  private external fun nativeRuntimeInfo(): String

  override fun definition() = ModuleDefinition {
    Name("HinsdaleEngine")
    AsyncFunction("analyze") { bytecodeHex: String, qualityTier: String, promise: Promise ->
      engineExecutor.execute {
        try { promise.resolve(nativeAnalyze(bytecodeHex, qualityTier)) }
        catch (error: Throwable) { promise.reject("ENGINE_FAILURE", "Embedded Hinsdale analysis failed: ${error.message ?: "unknown native error"}", error) }
      }
    }
    Function("runtimeInfo") { nativeRuntimeInfo() }
    OnDestroy { engineExecutor.shutdownNow() }
  }
}

File: modules/hinsdale-engine/ios/HinsdaleEngineModule.swift
import ExpoModulesCore

@_silgen_name("hins_analyze_enveloped_json")
private func hinsAnalyzeEnvelopedJson(_ hex: UnsafePointer<CChar>, _ tier: UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?

@_silgen_name("hins_free_str")
private func hinsFreeString(_ value: UnsafeMutablePointer<CChar>)

@_silgen_name("hins_mobile_runtime_info")
private func hinsMobileRuntimeInfo() -> UnsafePointer<CChar>?

public class HinsdaleEngineModule: Module {
  public func definition() -> ModuleDefinition {
    Name("HinsdaleEngine")
    AsyncFunction("analyze") { (bytecodeHex: String, qualityTier: String) -> String in
      return bytecodeHex.withCString { hexPointer in
        qualityTier.withCString { tierPointer in
          guard let response = hinsAnalyzeEnvelopedJson(hexPointer, tierPointer) else {
            return "{\"ok\":false,\"error\":{\"code\":\"ENGINE_FAILURE\",\"message\":\"Embedded engine returned no response.\"}}"
          }
          defer { hinsFreeString(response) }
          return String(cString: response)
        }
      }
    }
    Function("runtimeInfo") {
      guard let response = hinsMobileRuntimeInfo() else { return "{}" }
      return String(cString: response)
    }
  }
}
